File: frontend/src/services/ocrService.js
import Optiic from 'optiic';

// Initialize Optiic with API key
// You should store this in environment variables
const optiic = new Optiic({
  apiKey: import.meta.env.VITE_OPTIIC_API_KEY || 'your_api_key_here'
});

/**
 * Process an image file using Optiic OCR
 * @param {File} file - The image file to process
 * @returns {Promise} - Promise with OCR results
 */
export const processImageWithOptiic = async (file) => {
  try {
    // Convert file to base64 or use file path
    const result = await optiic.process({
      image: file
    });
    
    return result;
  } catch (error) {
    console.error('Optiic OCR Error:', error);
    throw error;
  }
};

/**
 * Process an image from URL using Optiic OCR
 * @param {string} url - The URL of the image to process
 * @returns {Promise} - Promise with OCR results
 */
export const processImageUrlWithOptiic = async (url) => {
  try {
    const result = await optiic.process({
      url: url
    });
    
    return result;
  } catch (error) {
    console.error('Optiic OCR URL Error:', error);
    throw error;
  }
};

/**
 * Parse Optiic OCR result into grocery items
 * @param {Object} ocrResult - Raw OCR result from Optiic
 * @returns {Array} - Array of parsed grocery items
 */
export const parseOCRResultToGroceryItems = (ocrResult) => {
  try {
    // Assuming Optiic returns structured text data
    const { text, lines, confidence } = ocrResult;
    
    // Parse the text to extract grocery items
    // This is a sample parser - adjust based on actual Optiic response format
    const items = [];
    
    if (lines && Array.isArray(lines)) {
      lines.forEach((line) => {
        // Try to extract item information from each line
        // Format: "Product Name Qty Price"
        const match = line.text.match(/^(.+?)\s+(\d+)\s+\$?(\d+\.?\d*)/);
        
        if (match) {
          items.push({
            name: match[1].trim(),
            quantity: parseInt(match[2]),
            price: parseFloat(match[3]),
            category: categorizeItem(match[1]),
            confidence: line.confidence || confidence
          });
        } else {
          // Try alternative pattern matching
          const itemMatch = extractItemFromLine(line.text);
          if (itemMatch) {
            items.push(itemMatch);
          }
        }
      });
    } else if (text) {
      // Fallback: Parse plain text
      const parsedItems = parseTextToItems(text);
      items.push(...parsedItems);
    }
    
    return items;
  } catch (error) {
    console.error('Error parsing OCR result:', error);
    return [];
  }
};

/**
 * Extract item information from a single line of text
 */
const extractItemFromLine = (lineText) => {
  // Multiple pattern matching strategies
  const patterns = [
    // Pattern 1: Name Quantity Price
    /^(.+?)\s+(\d+)\s+\$?(\d+\.?\d*)$/,
    // Pattern 2: Name - Quantity - Price
    /^(.+?)\s*-\s*(\d+)\s*-\s*\$?(\d+\.?\d*)$/,
    // Pattern 3: Name x Quantity @ Price
    /^(.+?)\s+x\s*(\d+)\s*@\s*\$?(\d+\.?\d*)$/,
    // Pattern 4: Quantity x Name @ Price
    /^(\d+)\s*x\s*(.+?)\s*@\s*\$?(\d+\.?\d*)$/
  ];
  
  for (const pattern of patterns) {
    const match = lineText.match(pattern);
    if (match) {
      // Check if quantity is in first or second position
      const isQuantityFirst = pattern.source.startsWith('^(\\d+)');
      
      return {
        name: isQuantityFirst ? match[2].trim() : match[1].trim(),
        quantity: isQuantityFirst ? parseInt(match[1]) : parseInt(match[2]),
        price: parseFloat(match[3]),
        category: categorizeItem(isQuantityFirst ? match[2] : match[1])
      };
    }
  }
  
  return null;
};

/**
 * Parse plain text to grocery items
 */
const parseTextToItems = (text) => {
  const items = [];
  const lines = text.split('\n');
  
  lines.forEach((line) => {
    const item = extractItemFromLine(line.trim());
    if (item) {
      items.push(item);
    }
  });
  
  return items;
};

/**
 * Categorize item based on name
 */
const categorizeItem = (itemName) => {
  const name = itemName.toLowerCase();
  
  // Dairy products
  if (name.includes('milk') || name.includes('cheese') || name.includes('yogurt') || 
      name.includes('butter') || name.includes('cream') || name.includes('egg')) {
    return 'Dairy';
  }
  
  // Bakery
  if (name.includes('bread') || name.includes('bun') || name.includes('cake') || 
      name.includes('pastry') || name.includes('cookie')) {
    return 'Bakery';
  }
  
  // Produce
  if (name.includes('tomato') || name.includes('potato') || name.includes('onion') || 
      name.includes('carrot') || name.includes('lettuce') || name.includes('fruit') ||
      name.includes('apple') || name.includes('banana') || name.includes('orange')) {
    return 'Produce';
  }
  
  // Meat
  if (name.includes('chicken') || name.includes('beef') || name.includes('pork') || 
      name.includes('fish') || name.includes('meat') || name.includes('sausage')) {
    return 'Meat';
  }
  
  // Beverages
  if (name.includes('juice') || name.includes('soda') || name.includes('water') || 
      name.includes('coffee') || name.includes('tea') || name.includes('drink')) {
    return 'Beverages';
  }
  
  // Snacks
  if (name.includes('chips') || name.includes('snack') || name.includes('candy') || 
      name.includes('chocolate')) {
    return 'Snacks';
  }
  
  // Default
  return 'Other';
};

export default {
  processImageWithOptiic,
  processImageUrlWithOptiic,
  parseOCRResultToGroceryItems
};

File: backend/app/utils/ocr_service.py
import pytesseract
from PIL import Image
import io
import re
from typing import List, Dict, Any
import logging
import os
import platform

logger = logging.getLogger(__name__)

# Configure tesseract path for Windows
if platform.system() == 'Windows':
    # Common installation paths
    possible_paths = [
        r'C:\Program Files\Tesseract-OCR\tesseract.exe',
        r'C:\Program Files (x86)\Tesseract-OCR\tesseract.exe',
    ]
    for path in possible_paths:
        if os.path.exists(path):
            pytesseract.pytesseract.tesseract_cmd = path
            logger.info(f"Tesseract found at: {path}")
            break

class OCRService:
    """
    OCR Service for extracting text and structured data from images
    Uses pytesseract for local OCR processing
    """
    
    @staticmethod
    async def process_image(image_bytes: bytes) -> Dict[str, Any]:
        """
        Process image bytes and extract text using OCR
        
        Args:
            image_bytes: Raw image bytes
            
        Returns:
            Dict containing extracted text and structured data
        """
        logger.info(f"Starting OCR processing on {len(image_bytes)} bytes")
        
        try:
            # Open image from bytes
            image = Image.open(io.BytesIO(image_bytes))
            logger.info(f"Image opened successfully: {image.size}, mode: {image.mode}")
            
            # Convert to RGB if necessary
            if image.mode != 'RGB':
                image = image.convert('RGB')
                logger.info("Converted image to RGB")
            
            # Extract text using pytesseract
            logger.info("Calling pytesseract.image_to_string...")
            text = pytesseract.image_to_string(image)
            logger.info(f"Extracted text length: {len(text)} characters")
            logger.info(f"Text preview: {text[:200]}")
            
            # Get detailed data with bounding boxes
            logger.info("Getting detailed OCR data...")
            data = pytesseract.image_to_data(image, output_type=pytesseract.Output.DICT)
            
            # Parse into structured format
            lines = OCRService._extract_lines(data)
            logger.info(f"Extracted {len(lines)} lines")
            
            confidence = OCRService._calculate_average_confidence(data)
            logger.info(f"Average confidence: {confidence}")
            
            return {
                'success': True,
                'text': text,
                'lines': lines,
                'confidence': confidence
            }
            
        except Exception as e:
            logger.error(f"OCR Processing Error: {str(e)}", exc_info=True)
            return {
                'success': False,
                'error': str(e),
                'text': '',
                'lines': [],
                'confidence': 0
            }
    
    @staticmethod
    def _extract_lines(data: Dict) -> List[Dict[str, Any]]:
        """Extract lines from OCR data"""
        lines = []
        n_boxes = len(data['text'])
        
        current_line = []
        current_block = -1
        
        for i in range(n_boxes):
            if int(data['conf'][i]) > 30:  # Filter low confidence results
                block_num = data['block_num'][i]
                
                if block_num != current_block and current_line:
                    # New block, save current line
                    line_text = ' '.join(current_line)
                    if line_text.strip():
                        lines.append({
                            'text': line_text.strip(),
                            'confidence': data['conf'][i]
                        })
                    current_line = []
                    current_block = block_num
                
                text = data['text'][i].strip()
                if text:
                    current_line.append(text)
        
        # Add last line
        if current_line:
            line_text = ' '.join(current_line)
            if line_text.strip():
                lines.append({
                    'text': line_text.strip(),
                    'confidence': 0
                })
        
        return lines
    
    @staticmethod
    def _calculate_average_confidence(data: Dict) -> float:
        """Calculate average confidence score"""
        confidences = [int(c) for c in data['conf'] if c != '-1']
        if confidences:
            return sum(confidences) / len(confidences)
        return 0.0
    
    @staticmethod
    def parse_grocery_items(ocr_result: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        Parse OCR result into grocery items
        
        Args:
            ocr_result: OCR result dict with text and lines
            
        Returns:
            List of parsed grocery items
        """
        items = []
        
        lines = ocr_result.get('lines', [])
        
        for line in lines:
            line_text = line.get('text', '')
            item = OCRService._parse_line_to_item(line_text)
            
            if item:
                item['confidence'] = line.get('confidence', 0)
                items.append(item)
        
        # If no items parsed from lines, try parsing full text
        if not items:
            text = ocr_result.get('text', '')
            items = OCRService._parse_text_to_items(text)
        
        return items
    
    @staticmethod
    def _parse_line_to_item(line_text: str) -> Dict[str, Any]:
        """
        Parse a single line of text to extract item information
        
        Supported patterns:
        - "Product Name Quantity Price"
        - "Product Name x Quantity @ Price"
        - "Quantity x Product Name @ Price"
        - "Product Name - Quantity - Price"
        """
        patterns = [
            # Pattern: Name Quantity Price (e.g., "Milk 10 3.99")
            r'^(.+?)\s+(\d+)\s+\$?(\d+\.?\d*)$',
            
            # Pattern: Name x Quantity @ Price (e.g., "Milk x 10 @ 3.99")
            r'^(.+?)\s+x\s*(\d+)\s*@\s*\$?(\d+\.?\d*)$',
            
            # Pattern: Quantity x Name @ Price (e.g., "10 x Milk @ 3.99")
            r'^(\d+)\s*x\s*(.+?)\s*@\s*\$?(\d+\.?\d*)$',
            
            # Pattern: Name - Quantity - Price (e.g., "Milk - 10 - 3.99")
            r'^(.+?)\s*-\s*(\d+)\s*-\s*\$?(\d+\.?\d*)$',
            
            # Pattern: Name Qty: Quantity Price: Price
            r'^(.+?)\s+[Qq]ty:?\s*(\d+)\s+[Pp]rice:?\s*\$?(\d+\.?\d*)$',
        ]
        
        for idx, pattern in enumerate(patterns):
            match = re.search(pattern, line_text.strip())
            if match:
                # Pattern 3 has quantity first
                is_qty_first = idx == 2
                
                name = match.group(2).strip() if is_qty_first else match.group(1).strip()
                quantity = int(match.group(1)) if is_qty_first else int(match.group(2))
                price = float(match.group(3))
                
                return {
                    'name': name,
                    'quantity': quantity,
                    'price': price,
                    'category': OCRService._categorize_item(name)
                }
        
        return None
    
    @staticmethod
    def _parse_text_to_items(text: str) -> List[Dict[str, Any]]:
        """Parse full text to extract items"""
        items = []
        lines = text.split('\n')
        
        for line in lines:
            item = OCRService._parse_line_to_item(line)
            if item:
                items.append(item)
        
        return items
    
    @staticmethod
    def _categorize_item(item_name: str) -> str:
        """Categorize item based on name keywords"""
        name_lower = item_name.lower()
        
        # Dairy products
        dairy_keywords = ['milk', 'cheese', 'yogurt', 'butter', 'cream', 'egg']
        if any(keyword in name_lower for keyword in dairy_keywords):
            return 'Dairy'
        
        # Bakery
        bakery_keywords = ['bread', 'bun', 'cake', 'pastry', 'cookie', 'biscuit']
        if any(keyword in name_lower for keyword in bakery_keywords):
            return 'Bakery'
        
        # Produce
        produce_keywords = ['tomato', 'potato', 'onion', 'carrot', 'lettuce', 'fruit',
                           'apple', 'banana', 'orange', 'vegetable', 'veggie']
        if any(keyword in name_lower for keyword in produce_keywords):
            return 'Produce'
        
        # Meat
        meat_keywords = ['chicken', 'beef', 'pork', 'fish', 'meat', 'sausage', 'bacon']
        if any(keyword in name_lower for keyword in meat_keywords):
            return 'Meat'
        
        # Beverages
        beverage_keywords = ['juice', 'soda', 'water', 'coffee', 'tea', 'drink', 'cola']
        if any(keyword in name_lower for keyword in beverage_keywords):
            return 'Beverages'
        
        # Snacks
        snack_keywords = ['chips', 'snack', 'candy', 'chocolate', 'crisp']
        if any(keyword in name_lower for keyword in snack_keywords):
            return 'Snacks'
        
        # Canned/Packaged
        packaged_keywords = ['can', 'jar', 'box', 'packet', 'pack']
        if any(keyword in name_lower for keyword in packaged_keywords):
            return 'Packaged'
        
        return 'Other'
